basis/math/matrices/matrices.factor

```diff
@@ -197,77 +297,316 @@ ERROR: negative-power-matrix m n ;
+: mmax ( m -- n ) [ -1/0. ] dip [ [ max ] each ] each ;
+: mnorm ( m -- m' ) dup mmax abs m/n ;
 
-: dimension-range ( matrix -- dim range )
-    dim [ matrix-coordinates ] [ first [1,b] ] bi ;
+! implementation details of slightly complicated math like gram schmidt
+<PRIVATE
+: (gram-schmidt) ( v seq -- newseq )
+    [ dupd proj v- ] each ;
 
-: upper-matrix-indices ( matrix -- matrix' )
-    dimension-range <reversed> [ tail-slice* >array ] 2map concat ;
+: (m^n) ( m n -- n )
+    make-bits over first length <identity-matrix>
+    [ [ dupd mdot ] when [ dup mdot ] dip ] reduce nip ;
+PRIVATE>
 
-: lower-matrix-indices ( matrix -- matrix' )
-    dimension-range [ head-slice >array ] 2map concat ;
+: gram-schmidt ( matrix -- orthogonal )
+    [ V{ } clone [ over (gram-schmidt) suffix! ] reduce ] keep like ;
+
+: gram-schmidt-normalize ( matrix -- orthonormal )
+    gram-schmidt [ normalize ] map ; inline
+
+DEFER: multiplicative-inverse
+! A^-1 is the inverse but other negative powers are nonsense
+: m^n ( m n -- n ) {
+        { [ dup -1 = ] [ drop multiplicative-inverse ] }
+        { [ dup 0 >= ] [ (m^n) ] }
+        [ negative-power-matrix ]
+    } cond ;
+
+: n^m ( n m -- n ) swap m^n ;
+
+: kronecker-product ( m1 m2 -- m )
+    '[ [ _ n*m  ] map ] map stitch stitch ;
+
+: outer-product ( u v -- m )
+    '[ _ n*v ] map ;
+
+! -------------------------------------------------
+! numerical analysis of matrices follows
+<PRIVATE
+: (rows-iota) ( matrix -- rows-iota )
+    dim first <iota> ;
+: (cols-iota) ( matrix -- cols-iota )
+    dim second <iota> ;
+
+: simple-rows-except ( matrix desc quot -- others )
+    curry [ dup (rows-iota) ] dip
+    pick reject-as swap rows ; inline
+
+: simple-cols-except ( matrix desc quot -- others )
+    curry [ dup (cols-iota) ] dip
+    pick reject-as swap cols transpose ; inline ! need to un-transpose the result of cols
+
+CONSTANT: scalar-except-quot [ = ]
+CONSTANT: sequence-except-quot [ member? ]
+
+: square-rank ( square-matrix -- rank ) ;
+: nonsquare-rank ( matrix -- rank ) ;
+PRIVATE>
+
+GENERIC: rank ( matrix -- rank )
+M: zero-matrix rank
+    drop +zero-rank+ ;
+
+M: square-matrix rank
+    square-rank ;
+
+M: matrix rank
+    nonsquare-rank ;
+
+GENERIC: nullity ( matrix -- nullity )
+
+:: >square-matrix ( m -- subset )
+    m dim first2 :> ( x y ) {
+        { [ x y = ] [ m ] }
+        { [ x y < ] [ x <iota> m cols transpose ] }
+        { [ x y > ] [ y <iota> m rows ] }
+    } cond ;
+
+! well-defined for square matrices; but works on nonsquare too
+: main-diagonal ( matrix -- seq )
+    >square-matrix [ swap nth ] map-index ; inline
+
+! top right to bottom left; reverse the result if you expected it to start in the lower left
+: anti-diagonal ( matrix -- seq )
+    >square-matrix [ swap nth-end ] map-index ; inline
+
+ALIAS: transpose flip
+
+! VERY slow implementation
+: anti-transpose ( matrix -- newmatrix )
+    [ reverse ] map transpose [ reverse ] map ;
+
```

Comment:
If it is so slow is there a use case for it?

Reply:
I finally committed the proper implementation that's just as fast as `flip`!